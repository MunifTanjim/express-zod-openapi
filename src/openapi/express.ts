import type { Application, Router } from 'express'
import type { RegisteredExpressOpenAPIPlugin } from '../plugins'
import type { HttpMethod } from '../types'
import type { OpenAPISpecification } from './specification'

const enum StackName {
  QUERY = 'query',
  EXPRESS_INIT = 'expressInit',
  BOUND_DISPATCH = 'bound dispatch',
  ROUTER = 'router',
  MOUNTED_APP = 'mounted_app',
}

interface Key {
  name: string | number
  prefix: string
  suffix: string
  pattern: string
  modifier: string
}

interface Layer {
  handle: Router | ((...params: any[]) => any)
  params?: Record<string, any>
  path?: string
  keys: Key[]
  regexp: RegExp & { fast_star?: boolean; fast_slash?: boolean }
}

interface StackLayer extends Layer {
  name: StackName
  route?: Route
}

interface RouteLayer extends Layer {
  name: string
  method: HttpMethod
}

export interface Route {
  path: string | string[]
  stack: RouteLayer[]
  methods: {
    [method in HttpMethod | '_all']?: boolean
  }
}

const regexPattern = {
  param: /\(\?:\(\[\^\\\/]\+\?\)\)/,
  path: /^\^\\\/(?:(:?[\w\\.-]*(?:\\\/:?[\w\\.-]*)*)|(\(\?:\(\[\^\\\/]\+\?\)\)))\\/,
  rootPath: /^\^\\\/\?\(\?=\\\/\|\$\)$/,
}

const getPathString = (
  pathRegex: RegExp,
  keys: Layer['keys'],
): string | null => {
  const pathRegexpSource = pathRegex.source

  let regexString = pathRegexpSource

  let regexStringExecArray = regexPattern.path.exec(regexString)

  let paramIndex = 0
  while (regexPattern.param.test(regexString)) {
    const param = `:${keys[paramIndex].name}`

    regexString = regexString.replace(regexPattern.param, param)

    paramIndex++
  }

  if (regexString !== pathRegexpSource) {
    regexStringExecArray = regexPattern.path.exec(regexString)
  }

  const pathString = regexStringExecArray?.[1].replace(/\\\//g, '/')

  return pathString ?? null
}

const processRoute = (
  spec: OpenAPISpecification,
  route: Route,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[],
): void => {
  const routePaths = Array.isArray(route.path) ? route.path : [route.path]

  routePaths.forEach((routePath) => {
    const path = `${
      basePath ? `${basePath}${routePath === '/' ? '' : routePath}` : routePath
    }`.replace(/:([a-zA-Z0-9_]+)/g, '{$1}')

    const methods = Object.keys(route.methods).filter(
      (method): method is HttpMethod => method !== '_all',
    )

    if (!spec.paths[path]) {
      spec.setPathItem(path, {})
    }

    for (const method of methods) {
      spec.setPathItemOperation(path, method, {
        responses: {
          default: {
            description: '',
          },
        },
      })

      for (const plugin of plugins) {
        const stashedValue = plugin.stash.find(route, method)
        if (stashedValue) {
          plugin.processRoute(plugin.specification, stashedValue, {
            path,
            method,
          })
        }
      }
    }
  })
}

export const processExpressRouters = (
  spec: OpenAPISpecification,
  app: Application | Router,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[],
): void => {
  const stack: StackLayer[] = app.stack ?? (app as Application)._router?.stack

  if (!stack) {
    spec.setPathItem(basePath, {})

    return
  }

  for (const layer of stack) {
    // terminal route
    if (layer.route) {
      processRoute(spec, layer.route, basePath, plugins)

      continue
    }

    if (
      layer.name !== StackName.ROUTER &&
      layer.name !== StackName.BOUND_DISPATCH &&
      layer.name !== StackName.MOUNTED_APP
    ) {
      continue
    }

    if (regexPattern.path.test(layer.regexp.source)) {
      const parsedPath = getPathString(layer.regexp, layer.keys)

      if (parsedPath) {
        processExpressRouters(
          spec,
          layer.handle as Router,
          basePath + '/' + parsedPath,
          plugins,
        )
      }

      continue
    }

    if (
      !layer.path &&
      layer.regexp &&
      !regexPattern.rootPath.test(layer.regexp.source)
    ) {
      const regexPath = ' RegExp(' + layer.regexp + ') '

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + regexPath,
        plugins,
      )

      continue
    }

    processExpressRouters(spec, layer.handle as Router, basePath, plugins)
  }
}
