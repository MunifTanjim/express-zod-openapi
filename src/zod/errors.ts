import type { ZodError } from 'zod'
import type { RequestSegment, ResponseSegment } from '../plugins/types'

export class RequestValidationError extends Error {
  segment: RequestSegment
  validationError: ZodError

  constructor(validationError: ZodError, segment: RequestSegment) {
    super(validationError.message)

    this.segment = segment
    this.validationError = validationError
  }
}

export class ResponseValidationError extends Error {
  segment: ResponseSegment
  validationError: ZodError

  constructor(validationError: ZodError, segment: ResponseSegment) {
    super(validationError.message)

    this.segment = segment
    this.validationError = validationError
  }
}
