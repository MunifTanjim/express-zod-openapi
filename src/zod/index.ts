export { zodToJsonSchema } from 'zod-to-json-schema'
export { RequestValidationError, ResponseValidationError } from './error'
