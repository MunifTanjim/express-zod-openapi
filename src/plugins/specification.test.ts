import express, { Application, ErrorRequestHandler, Handler } from 'express'
import request from 'supertest'
import { z } from 'zod'
import {
  GetSpecificationMiddleware,
  getSpecificationPlugin,
  RequestValidationError,
  ResponseValidationError,
} from '.'
import { ExpressOpenAPI } from '../index'

const preprocessors = {
  numberString: (v: unknown) => {
    const val = Number(v)
    return Number.isNaN(val) ? v : val
  },
}

describe('getSpecificationPlugin', () => {
  test('returns plugin', () => {
    const plugin = getSpecificationPlugin()
    expect(plugin.name).toMatchInlineSnapshot(`"zod-openapi-spec"`)
    expect(plugin.getMiddleware).toBeDefined()
    expect(plugin.processRoute).toBeDefined()
  })

  describe('request validation', () => {
    let spec: GetSpecificationMiddleware
    let app: Application

    const handler: Handler = (_req, res) => {
      res.status(200).json({
        pong: 42,
      })
    }

    const errorHandler: ErrorRequestHandler = (err, _req, res, _next) => {
      if (err instanceof RequestValidationError) {
        res.status(400).json({
          error: {
            message: err.message,
            segment: err.segment,
            // details: err.validationError.details.map(
            //   ({ message, path, type }) => ({ message, path, type })
            // ),
          },
        })

        return
      }

      res.status(500).json({
        error: {
          message: 'Server Error',
        },
      })
    }

    describe('segment validation', () => {
      beforeEach(() => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const validatorPlugin = getSpecificationPlugin({
          res: {
            skipValidation: true,
          },
        })

        spec = expressOpenApi.registerPlugin(validatorPlugin)
      })

      test('validates headers', async () => {
        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            req: {
              headers: z.object({
                authorization: z.string().regex(/^Bearer .+$/),
              }),
            },
          }),
          handler,
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .set('authorization', `JWT 42`)
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const validResponse1 = await request(app)
          .post('/ping')
          .set('authorization', `Bearer 42`)
        expect(validResponse1.status).toBe(200)
      })

      test('validates params', async () => {
        app.post(
          '/ping/:id',
          spec({
            operationId: 'ping',
            req: {
              params: z.object({
                id: z.preprocess(preprocessors.numberString, z.number().int()),
              }),
            },
          }),
          handler,
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app).post('/ping/forty-two')
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const validResponse1 = await request(app).post('/ping/42')
        expect(validResponse1.status).toBe(200)
      })

      test('validates query', async () => {
        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            req: {
              query: z.object({
                count: z.preprocess(
                  preprocessors.numberString,
                  z.number().int(),
                ),
              }),
            },
          }),
          handler,
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app).post('/ping')
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const invalidResponse2 = await request(app).post(
          '/ping?count=forty-two',
        )
        expect(invalidResponse2.status).toBe(400)
        expect(invalidResponse2.body).toMatchSnapshot()

        const validResponse1 = await request(app).post('/ping?count=42')
        expect(validResponse1.status).toBe(200)
      })

      test('validates body', async () => {
        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            req: {
              body: z.object({
                type: z.enum(['plastic', 'rubber', 'wood']).default('rubber'),
              }),
            },
          }),
          handler,
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .send({ type: 'paper' })
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const validResponse1 = await request(app).post('/ping')
        expect(validResponse1.status).toBe(200)

        const validResponse2 = await request(app)
          .post('/ping')
          .send({ type: 'wood' })
        expect(validResponse2.status).toBe(200)
      })
    })

    describe('value coercion', () => {
      const mockFn = jest.fn()

      beforeEach(() => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const validatorPlugin = getSpecificationPlugin({
          res: {
            skipValidation: true,
          },
        })

        spec = expressOpenApi.registerPlugin(validatorPlugin)

        mockFn.mockReset()
      })

      test('coerces w/o option debug/warnings', async () => {
        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            req: {
              body: z.object({
                count: z.preprocess(
                  preprocessors.numberString,
                  z.number().int(),
                ),
              }),
            },
          }),
          (req, _res, next) => {
            mockFn(req.body)
            next()
          },
          handler,
        )

        app.use(errorHandler)

        const response = await request(app).post('/ping').send({ count: '42' })
        expect(response.status).toBe(200)
        expect(mockFn.mock.calls).toMatchInlineSnapshot(`
            Array [
              Array [
                Object {
                  "count": 42,
                },
              ],
            ]
          `)
      })
    })

    describe('options: segmentOrder', () => {
      test('(default) validates query before body', async () => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const validatorPlugin = getSpecificationPlugin({
          res: {
            skipValidation: true,
          },
        })

        spec = expressOpenApi.registerPlugin(validatorPlugin)

        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            req: {
              body: z.object({
                type: z.enum(['plastic', 'rubber', 'wood']).default('rubber'),
              }),
              query: z.object({
                count: z.preprocess(
                  preprocessors.numberString,
                  z.number().int(),
                ),
              }),
            },
          }),
          handler,
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .send({ type: 'paper' })
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const invalidResponse2 = await request(app)
          .post('/ping?count=42')
          .send({ type: 'paper' })
        expect(invalidResponse2.status).toBe(400)
        expect(invalidResponse2.body).toMatchSnapshot()

        const validResponse1 = await request(app)
          .post('/ping?count=42')
          .send({ type: 'wood' })
        expect(validResponse1.status).toBe(200)
      })

      test('(custom) validates body before query', async () => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const validatorPlugin = getSpecificationPlugin({
          res: {
            skipValidation: true,
          },
          req: {
            segmentOrder: ['body', 'query'],
          },
        })

        spec = expressOpenApi.registerPlugin(validatorPlugin)

        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            req: {
              body: z.object({
                type: z.enum(['plastic', 'rubber', 'wood']).default('rubber'),
              }),
              query: z.object({
                count: z.preprocess(
                  preprocessors.numberString,
                  z.number().int(),
                ),
              }),
            },
          }),
          handler,
        )

        app.use(errorHandler)

        const invalidResponse1 = await request(app)
          .post('/ping')
          .send({ type: 'paper' })
        expect(invalidResponse1.status).toBe(400)
        expect(invalidResponse1.body).toMatchSnapshot()

        const invalidResponse2 = await request(app)
          .post('/ping')
          .send({ type: 'wood' })
        expect(invalidResponse2.status).toBe(400)
        expect(invalidResponse2.body).toMatchSnapshot()

        const validResponse1 = await request(app)
          .post('/ping?count=42')
          .send({ type: 'wood' })
        expect(validResponse1.status).toBe(200)
      })
    })
  })

  describe('response validation', () => {
    let spec: GetSpecificationMiddleware
    let app: Application

    const errorHandler = jest.fn((err, _req, res, _next) => {
      if (err instanceof ResponseValidationError) {
        res.status(500).json({
          error: {
            message: err.message,
            segment: err.segment,
            // details: err.validationError.details.map(
            //   ({ message, path, type }) => ({ message, path, type })
            // ),
          },
        })

        return
      }

      res.status(500).json({
        error: {
          message: 'Server Error',
        },
      })
    })

    describe('segment validation', () => {
      beforeEach(() => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const validatorPlugin = getSpecificationPlugin({
          req: {
            skipValidation: true,
          },
        })

        spec = expressOpenApi.registerPlugin(validatorPlugin)
      })

      test('validates headers', async () => {
        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            res: {
              200: {
                headers: z.object({
                  'x-count': z.preprocess(
                    preprocessors.numberString,
                    z.number().int(),
                  ),
                }),
              },
              default: {},
            },
          }),
          (req, res) => {
            const pass = Number(req.query.pass)

            if (pass) {
              res.set('x-count', '42')
            }

            res.status(200).json({
              pong: 42,
            })
          },
        )

        app.use(errorHandler)

        const responseFail = await request(app).post('/ping?pass=0')

        expect(responseFail.status).toBe(500)
        expect(responseFail.body).toMatchSnapshot()

        const responsePass = await request(app).post('/ping?pass=1')

        expect(responsePass.status).toBe(200)
      })

      test('validates body', async () => {
        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            res: {
              200: {
                body: z.object({
                  pong: z.number().int(),
                }),
              },
              default: {},
            },
          }),
          (req, res) => {
            const pass = Number(req.query.pass)

            res.status(200).json({
              pong: pass ? 42 : 'forty-two',
            })
          },
        )

        app.use(errorHandler)

        const responseFail = await request(app).post('/ping?pass=0')

        expect(responseFail.status).toBe(500)
        expect(responseFail.body).toMatchSnapshot()

        const responsePass = await request(app).post('/ping?pass=1')

        expect(responsePass.status).toBe(200)
      })

      test('works for res.send inside callback', async () => {
        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            res: {
              200: {
                body: z.object({
                  pong: z.number().int(),
                }),
              },
              default: {},
            },
          }),
          (req, res) => {
            const pass = Number(req.query.pass)

            setTimeout(() => {
              res.status(200).json({
                pong: pass ? 42 : 'forty-two',
              })
            }, 0)
          },
        )

        app.use(errorHandler)

        const responseFail = await request(app).post('/ping?pass=0')

        expect(responseFail.status).toBe(500)
        expect(responseFail.body).toMatchSnapshot()

        const responsePass = await request(app).post('/ping?pass=1')

        expect(responsePass.status).toBe(200)
      })
    })

    describe('options: res.skipValidation', () => {
      test('works as expected', async () => {
        const expressOpenApi = new ExpressOpenAPI()

        app = express()
        app.use(express.json())

        const specificationPlugin = getSpecificationPlugin({
          req: {
            skipValidation: true,
          },
          res: {
            skipValidation: true,
          },
        })

        spec = expressOpenApi.registerPlugin(specificationPlugin)

        app.post(
          '/ping',
          spec({
            operationId: 'ping',
            res: {
              200: {
                headers: z.object({
                  'x-count': z.number().int(),
                }),
              },
              default: {},
            },
          }),
          (req, res) => {
            const pass = Number(req.query.pass)

            if (pass) {
              res.set('x-count', '42')
            }

            res.status(200).json({
              pong: 42,
            })
          },
        )

        app.use(errorHandler)

        const response = await request(app).post('/ping?pass=0')

        expect(response.status).toBe(200)
      })
    })
  })
})
