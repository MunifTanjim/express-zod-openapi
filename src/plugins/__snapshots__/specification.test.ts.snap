// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSpecificationPlugin request validation options: segmentOrder (custom) validates body before query 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_enum_value",
    "options": [
      "plastic",
      "rubber",
      "wood"
    ],
    "path": [
      "type"
    ],
    "message": "Invalid enum value. Expected 'plastic' | 'rubber' | 'wood', received 'paper'"
  }
]",
    "segment": "body",
  },
}
`;

exports[`getSpecificationPlugin request validation options: segmentOrder (custom) validates body before query 2`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "count"
    ],
    "message": "Required"
  }
]",
    "segment": "query",
  },
}
`;

exports[`getSpecificationPlugin request validation options: segmentOrder (default) validates query before body 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "count"
    ],
    "message": "Required"
  }
]",
    "segment": "query",
  },
}
`;

exports[`getSpecificationPlugin request validation options: segmentOrder (default) validates query before body 2`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_enum_value",
    "options": [
      "plastic",
      "rubber",
      "wood"
    ],
    "path": [
      "type"
    ],
    "message": "Invalid enum value. Expected 'plastic' | 'rubber' | 'wood', received 'paper'"
  }
]",
    "segment": "body",
  },
}
`;

exports[`getSpecificationPlugin request validation segment validation validates body 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_enum_value",
    "options": [
      "plastic",
      "rubber",
      "wood"
    ],
    "path": [
      "type"
    ],
    "message": "Invalid enum value. Expected 'plastic' | 'rubber' | 'wood', received 'paper'"
  }
]",
    "segment": "body",
  },
}
`;

exports[`getSpecificationPlugin request validation segment validation validates headers 1`] = `
{
  "error": {
    "message": "[
  {
    "validation": "regex",
    "code": "invalid_string",
    "message": "Invalid",
    "path": [
      "authorization"
    ]
  }
]",
    "segment": "headers",
  },
}
`;

exports[`getSpecificationPlugin request validation segment validation validates params 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "id"
    ],
    "message": "Expected number, received string"
  }
]",
    "segment": "params",
  },
}
`;

exports[`getSpecificationPlugin request validation segment validation validates query 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "count"
    ],
    "message": "Required"
  }
]",
    "segment": "query",
  },
}
`;

exports[`getSpecificationPlugin request validation segment validation validates query 2`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "count"
    ],
    "message": "Expected number, received string"
  }
]",
    "segment": "query",
  },
}
`;

exports[`getSpecificationPlugin response validation segment validation validates body 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "pong"
    ],
    "message": "Expected number, received string"
  }
]",
    "segment": "body",
  },
}
`;

exports[`getSpecificationPlugin response validation segment validation validates headers 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "x-count"
    ],
    "message": "Required"
  }
]",
    "segment": "headers",
  },
}
`;

exports[`getSpecificationPlugin response validation segment validation works for res.send inside callback 1`] = `
{
  "error": {
    "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "pong"
    ],
    "message": "Expected number, received string"
  }
]",
    "segment": "body",
  },
}
`;
